backend/store/etcd/health_store.go

```diff
@@ -22,72 +25,100 @@ func isEmbeddedClient(clientURLs []string) bool {
+	healthResponse := &corev2.HealthResponse{}
 
 	// Do a get op against every cluster member. Collect the  memberIDs and
 	// op errors into a response map, and return this map as etcd health
 	// information.
 	mList, err := cluster.MemberList(ctx)
 	if err != nil {
-		logger.WithError(err).Warning("could not get the cluster member list")
+		logger.WithError(err).Error("could not get the cluster member list")
+		healthResponse.ClusterHealth = []*corev2.ClusterHealth{&corev2.ClusterHealth{
+			Name: "etcd client",
+			Err:  fmt.Sprintf("error getting cluster member list: %s", err.Error()),
+		}}
 		return healthResponse
 	}
+	logger.WithField("members", mList.Members).Info("retrieved cluster members")
 	healthResponse.Header = mList.Header
 
-	for _, member := range mList.Members {
-		health := &types.ClusterHealth{
-			MemberID: member.ID,
-			Name:     member.Name,
-		}
+	healths := make(chan *corev2.ClusterHealth, len(mList.Members))
+	var wg sync.WaitGroup
+	wg.Add(len(mList.Members))
 
-		var cli *clientv3.Client
-		var cliErr error
-
-		if isEmbeddedClient(member.ClientURLs) {
-			cli = s.client
-		} else {
-			cli, cliErr = clientv3.New(clientv3.Config{
-				Endpoints:   member.ClientURLs,
-				DialTimeout: 5 * time.Second,
-				TLS:         etcdClientTLSConfig,
-			})
-		}
+	go func() {
+		wg.Wait()
+		close(healths)
+	}()
 
-		if cliErr != nil {
-			logger.WithField("member", member.ID).WithError(cliErr).Error("unhealthy cluster member")
-			health.Err = cliErr.Error()
-			health.Healthy = false
-			healthResponse.ClusterHealth = append(healthResponse.ClusterHealth, health)
-			continue
-		}
-		defer func() {
-			_ = cli.Close()
-		}()
-
-		// Specify the client request timeout to get the health
-		ctx, cancel := context.WithTimeout(ctx, 5*time.Second)
-		_, getErr := cli.Get(ctx, "health")
-		cancel()
-
-		if getErr == nil || getErr == rpctypes.ErrPermissionDenied {
-			health.Err = ""
-			health.Healthy = true
-		} else {
-			health.Err = getErr.Error()
-			health.Healthy = false
-		}
+	for _, member := range mList.Members {
+		go func(id uint64, name string, urls []string) {
+			defer wg.Done()
+			select {
+			case healths <- s.getHealth(ctx, id, name, urls, etcdClientTLSConfig):
+			case <-ctx.Done():
+			}
+		}(member.ID, member.Name, member.ClientURLs)
+	}
 
+	for health := range healths {
+		logger.WithField("health", health).Info("cluster member health")
 		healthResponse.ClusterHealth = append(healthResponse.ClusterHealth, health)
 	}
 
-	alarmResponse, err := s.client.Maintenance.AlarmList(ctx)
+	sort.Slice(healthResponse.ClusterHealth, func(i, j int) bool {
+		return healthResponse.ClusterHealth[i].Name < healthResponse.ClusterHealth[j].Name
+	})
 
+	alarmResponse, err := s.client.Maintenance.AlarmList(ctx)
 	if err != nil {
 		logger.WithError(err).Error("failed to fetch etcd alarm list")
 	}
 
+	logger.WithField("alarms", len(alarmResponse.Alarms)).Info("cluster alarms")
```

Comment:
This will cause a panic if `alarmResponse` is nil because we had an error on line 115, since we only print a log entry and do not return.